### PR TITLE
[wip] Overhaul MessageList

### PR DESCRIPTION
**Pull request by Jikstra**
# ToDo

- [x] Fix https://github.com/deltachat/deltachat-desktop/issues/2149
- [x] Fix https://github.com/deltachat/deltachat-desktop/issues/2069
- [x] Fix https://github.com/deltachat/deltachat-desktop/issues/1684
- [x] Fix https://github.com/deltachat/deltachat-desktop/issues/1330
- [x] Fix pages getting loaded twice
- [x] Fix jump to message
- [x] msg change
  - [x] scroll restore buggy
- [x] delete message
- [x] send message
  - [x] not scrolling to send message
- [x] incoming message
  - [x] bugs
- [x] contact requests
- [x] mark all messages read when click jump-to-bottom-button
- [x] weird bug rendering empty messages or wrong chat
- [ ] Fix https://github.com/deltachat/deltachat-desktop/issues/1540
- [ ] Fix https://github.com/deltachat/deltachat-desktop/issues/1407
- [ ] Fix https://github.com/deltachat/deltachat-desktop/issues/1293
- [ ] highlight message on jumpToMessage
- [x] add missing scrollToBottom icon
- [x] show scrollToBottom button also when one page away from newest message
- [x] dont pop/remove markerOne when the new messages will be in view
- [ ] Fix https://github.com/deltachat/deltachat-desktop/issues/1483
- [x] Fix contact request list
- [x] Fix context menu
- [x] Fix chat audit log
- [x] Fix author name on quote

### CODE REVIEW

File: scss/message/_message-list.scss [hunk 1]
```diff
@@ -9,6 +9,45 @@
   background-size: cover;
   background-color: var(--chatViewBg);
   overflow: hidden;
+
+  .unread-message-counter {
+    position: relative;
+    left: calc(100% - 40px - 10px);
+    top: -(30px + 40px + 20px);
```
Comment (Simon-Laux) [inline scss/message/_message-list.scss:16]: scss vars would make sense here

File: scss/message/_message.scss [hunk 2]
```diff
@@ -373,6 +375,7 @@
   padding: 0 4px;
   padding-left: 9px;
   margin: 2px 0 5px 0;
+  height: 44px;
```
Comment (Simon-Laux) [inline scss/message/_message.scss:378]: ?

File: scss/message/_message.scss [hunk 3]
```diff
@@ -388,9 +391,8 @@
 
   p {
     font-size: 12px;
-    line-height: 12px;
     margin: 0;
-    max-height: 40px;
+    height: 30px;
```
Comment (Simon-Laux) [inline scss/message/_message.scss:395]: ?

File: src/main/deltachat/chatlist.ts [hunk 4]
```diff
@@ -230,7 +219,7 @@ export default class DCChatList extends SplitOut {
   }
 }
 // section: Internal functions
-
+// TODO: Delete
```
Comment (Simon-Laux) [inline src/main/deltachat/chatlist.ts:222]: ?

File: src/main/deltachat/messagelist.ts [hunk 5]
```diff
@@ -12,11 +15,17 @@ import {
...
+  Message2,
+  MessageQuote,
 } from '../../shared/shared-types'
 
 import { writeFile } from 'fs-extra'
 import tempy from 'tempy'
+import { getAccountsPath } from '../application-constants'
+import { join } from 'path'
+import { MessageType2 } from '../../shared/shared'
+import { MultiGrid } from 'react-virtualized'
```
Comment (Simon-Laux) [inline src/main/deltachat/messagelist.ts:28]: wtf

File: src/renderer/components/message/MessageList.tsx [hunk 6]
```diff
@@ -1,270 +1,616 @@
...
+  }: {
+	chat: ChatStoreState
+	refComposer: todo
+  }) {
+	const messageListRef = useRef(null)
+	const messageListWrapperRef = useRef(null)
+	const messageListTopRef = useRef(null)
+	const messageListBottomRef = useRef(null)
+	const onMessageListStoreEffect = (action: Action) => {
+	  if (action.type === 'SCROLL_BEFORE_LAST_PAGE') {
```
Comment (Simon-Laux) [inline src/renderer/components/message/MessageList.tsx:131]: convert type to typescript enum

File: src/renderer/components/message/MessageWrapper.tsx [hunk 7]
```diff
@@ -1,17 +1,48 @@
...
+  isDeviceChat: boolean 
+  unreadMessageInViewIntersectionObserver: React.MutableRefObject<any>
 }
-
+     
 export const MessageWrapper = (props: RenderMessageProps) => {
+  const state = props.message.msg.state
+  const shouldInViewObserve = state === C.DC_STATE_IN_FRESH || state === C.DC_STATE_IN_NOTICED
+
+  useLayoutEffect(() => {
```
Comment (Simon-Laux) [inline src/renderer/components/message/MessageWrapper.tsx:21]: please add more comments what this does

File: src/renderer/ipc.ts [hunk 8]
```diff
@@ -76,15 +76,15 @@ function callDcMethod(
...
       'Got back return: [truncated]'
-    )
-    /*log.debug(
+    )*/
+    log.debug(
       `EVENT_DD_DISPATCH_RETURN_${identifier}_${methodName}`,
       'Got back return: [',
       returnValue
-    )*/
+    )
```
Comment (Simon-Laux) [inline src/renderer/ipc.ts:87]: Uncomment before merge again

File: src/renderer/stores/store.ts [hunk 9]
```diff
@@ -50,17 +50,22 @@ export class Store<S> {
...
       this.log.debug(`DISPATCHING of "${action.type}" changed the state.`)
       this.state = state
       this.listeners.forEach(listener => listener(this.state))
     }
   }
+  
+  dispatch2(type: string, payload: any, id: number) {
+    const action: Action = { type, payload, id}
+    return this.dispatch(action)
+  }
```
Comment (Simon-Laux) [inline src/renderer/stores/store.ts:68]: Is this still used?

File: src/renderer/stores/store2.ts [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+  if (args.currentlyDispatchedCounter !== 0) return false
+  return true
+}
+
+export function OnlySetStateIfIncrementingDispatchedCounterDidntIncrease(args: BeforeSetStateParameters) {
+  if (args.incrementingDispatchedCounter !== args.yourIncrementingDispatchedCounter) return false
+  return true
+}
+
+export class Store<S> {
```
Comment (Simon-Laux) [inline src/renderer/stores/store2.ts:61]: add comment explaining what this store makes different

File: src/shared/shared-types.d.ts [hunk 11]
```diff
@@ -154,22 +156,66 @@ export interface MessageTypeAttachment {
   fileSize: string
 }
 
-export type msgStatus =
-  | 'error'
+export type MessageState = C.DC_STATE_IN_FRESH
```
Comment (Simon-Laux) [inline src/shared/shared-types.d.ts:159]: please use enum instead:
like:
```ts
export const enum MessageState2 {
  IN_FRESH = C.DC_STATE_IN_FRESH,
  IN_NOTICED = C.DC_STATE_IN_NOTICED,
  OUT_DELIVERED = C.DC_STATE_OUT_DELIVERED,
  OUT_DRAFT = C.DC_STATE_OUT_DRAFT,
  OUT_FAILED = C.DC_STATE_OUT_FAILED,
  OUT_MDN_RCVD = C.DC_STATE_OUT_MDN_RCVD,
  OUT_PENDING = C.DC_STATE_OUT_PENDING,
  OUT_PREPARING = C.DC_STATE_OUT_PREPARING,
}
```

File: CHANGELOG.md [hunk 12]
```diff
@@ -67,8 +67,8 @@
 
 ### Changed
 
-- update translations (22.05.2021)
-- Update deltachat-node to `v1.55.0`
+- update translations (02.05.2021)
+- Update deltachat-node to `v1.54.0`
```
Comment (Simon-Laux) [inline CHANGELOG.md:71]: this change is unrelated and wrong, probably a mistake that got in while rebasing

File: src/renderer/ipc.ts (hunk 8 above)
Comment (Jikstra) [reply src/renderer/ipc.ts:87]: Or let's leave it in for debugging?

Comment (Simon-Laux) [reply src/renderer/ipc.ts:87]: you can put it behind an if statement if its a method that you wanna debug, but I don't see the point into logging everything. especially as this makes the logs contain more sensitive information.
